torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2899,10 +2901,20 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     ), "Currently the way for _is_sharded to be False is \
                         world_size == 1 or sharding_stratagy is set to be NO_SHARD"
                     if self.sharding_strategy == ShardingStrategy.NO_SHARD:
-                        dist.all_reduce(param.grad, group=self.process_group)
-                        if self.gradient_postdivide_factor > 1:
-                            # Average grad by world_size for consistency with PyTorch DDP.
-                            param.grad.div_(self.gradient_postdivide_factor)
+                        # if a communication hook was registered,
+                        # then a model has attribute ``communication_hook``
+                        if hasattr(self, "communication_hook"):
```

Comment:
setting `self.comm_hook = None` in ctor might be cleaner, as we can just check `if self.comm_hook is not None` instead of hasattr check.

Reply:
Got it. Adjusted logic in the new commit.